SYCL/ESIMD/esimd_test_utils.hpp

```diff
@@ -99,26 +100,73 @@ bool write_binary_file(const char *fname, const std::vector<T> &vec,
 }
 
 template <typename T>
-bool cmp_binary_files(const char *fname1, const char *fname2, T tolerance) {
+bool cmp_binary_files(const char *fname1, const char *fname2, const T tolerance,
+                      const double toleratedMismatchRate = 0,
+                      const int mismatchReportThrottleLimit = 5) {
+
+  if (toleratedMismatchRate) {
+    if (toleratedMismatchRate >= 1 || toleratedMismatchRate < 0) {
+      std::cerr << "Tolerated mismatch rate (" << toleratedMismatchRate
+                << ") must be set within [0, 1) range" << std::endl;
+      return false;
+    }
+
+    std::cerr << "Tolerated mismatch rate set to " << toleratedMismatchRate
+              << std::endl;
+  }
+
   const auto vec1 = read_binary_file<T>(fname1);
   const auto vec2 = read_binary_file<T>(fname2);
+
   if (vec1.size() != vec2.size()) {
     std::cerr << fname1 << " size is " << vec1.size();
     std::cerr << " whereas " << fname2 << " size is " << vec2.size()
               << std::endl;
     return false;
   }
-  for (size_t i = 0; i < vec1.size(); i++) {
+
+  double totalMismatches = 0;
+  const double size = vec1.size();
+  for (size_t i = 0; i < size; i++) {
     if (abs(vec1[i] - vec2[i]) > tolerance) {
-      std::cerr << "Mismatch at " << i << ' ';
-      if (sizeof(T) == 1) {
-        std::cerr << (int)vec1[i] << " vs " << (int)vec2[i] << std::endl;
-      } else {
-        std::cerr << vec1[i] << " vs " << vec2[i] << std::endl;
+      if (!toleratedMismatchRate ||
+          (totalMismatches < mismatchReportThrottleLimit)) {
+
+        std::cerr << "Mismatch at " << i << ' ';
+        if (sizeof(T) == 1) {
+          std::cerr << (int)vec1[i] << " vs " << (int)vec2[i];
+        } else {
+          std::cerr << vec1[i] << " vs " << vec2[i];
+        }
+
+        if (!toleratedMismatchRate) {
+          std::cerr << std::endl;
+          return false;
+        } else {
+          std::cerr << ". Current mismatch rate: " << std::setprecision(8)
+                    << std::fixed << totalMismatches / size << std::endl;
```

Comment:
```suggestion
                    << std::fixed << totalMismatches / (double)size << std::endl;
```
accompanies the above suggestion to make `size` of `size_t` type.

Reply:
TBH, I was relying on implicit conversions.
Yes, implicit conversions can be evil (int promotion is just one to name among those, not related here 
however).

If you explicitly oppose any implicit conversions, I can change. If not, let's leave it as is, not to make more code.

Reply:
if size becomes size_t as I suggested above, then totalMismatches / size would be 0

Reply:
totalMismatches is of type double, so totalMismatches / size will not be 0 unless totalMismatches is 0.0.
"size" will be implicitly converted to double in this expression.

Reply:
totalMismatches should be integral as well then. I did not even expect it to be double, very counter-intuitive. Please make them both integral and do casting where needed.

Reply:
The type of totalMismatches conforms to its use case (also, intuition is subjective).

Reply:
please change `totalMismatches` and `size` to integer type to avoid hard-to-detect conversion errors when double can't accurately represent these integer (by nature) values.

Reply:
The problem would be that at some point (very big value) we'd stop incrementing.

Reply:
For double after reaching 2^53 we'd stop incrementing the value if incrementing by 1, that's true.
If there'll ever be a > petabyte file checked in llvm-test-suite, well, yes, that could be a problem.

Reply:
Changed.